File: backend/app/services/scoring_service.py
from typing import List, Dict, Any, Tuple
from app.models.student import StudentProfile

class ScoringService:
    @staticmethod
    def calculate_score(
        item: Dict[str, Any],
        student_profile: StudentProfile
    ) -> Tuple[float, str, List[str]]:
        """Calculates multi-factor recommendation score (0-100), label, and match reasons."""
        stopwords = {"i", "want", "to", "become", "a", "an", "the", "in", "and", "have", "know", "per", "week", "hours", "my", "is", "for", "with", "get"}
        raw_words = [k.strip().lower() for k in student_profile.goal.split() if k.strip()]
        goal_keywords = [w for w in raw_words if w not in stopwords]
        if not goal_keywords:
            goal_keywords = raw_words or ["ai", "engineer"]

        # Item text
        item_text = (
            f"{item.get('name', '')} {item.get('title', '')} {item.get('description', '')} "
            f"{item.get('skills', '')} {item.get('department', '')} {item.get('category', '')}"
        ).lower()

        # 1. Goal Match (55%)
        goal_hits = sum(1 for kw in goal_keywords if kw in item_text)
        if goal_hits > 0:
            goal_match = min(100.0, 50.0 + (goal_hits * 25.0))
        else:
            goal_match = 10.0

        # 2. Skill Match (20%)
        student_skill_names = [s.name.lower() for s in student_profile.skills]
        item_skills = [s.strip().lower() for s in str(item.get('skills', '')).split('|') if s.strip()]
        
        if not item_skills:
            skill_match = 50.0
        else:
            shared_skills = set(student_skill_names).intersection(set(item_skills))
            skill_match = min(100.0, (len(shared_skills) / max(1, len(item_skills))) * 80.0 + 20.0) if shared_skills else 40.0

        # 3. Time Fit (15%)
        item_hours = float(item.get('hours_per_week', 2))
        avail_hours = student_profile.available_hours_per_week
        if item_hours <= avail_hours:
            time_fit = 100.0
        else:
            over = item_hours - avail_hours
            time_fit = max(10.0, 100.0 - (over * 25.0))

        # 4. Opportunity Value (10%)
        item_type = str(item.get('type', '')).lower()
        if "research" in item_type or "fellowship" in item_type or "hackathon" in item_type or "internship" in item_type:
            opportunity_value = 95.0
        elif "course" in item_type or "workshop" in item_type:
            opportunity_value = 85.0
        else:
            opportunity_value = 75.0

        # Weighted Total Score
        total_score = (goal_match * 0.55) + (skill_match * 0.20) + (time_fit * 0.15) + (opportunity_value * 0.10)
        total_score = round(min(100.0, max(0.0, total_score)), 1)

        # Score Label
        if total_score >= 90:
            label = "Excellent Match"
        elif total_score >= 75:
            label = "Strong Match"
        elif total_score >= 60:
            label = "Good Match"
        elif total_score >= 40:
            label = "Possible Match"
        else:
            label = "Low Match"

        # Match Reasons
        reasons = []
        if goal_match >= 70:
            reasons.append(f"Directly aligns with your '{student_profile.goal}' goal")
        if item_hours <= avail_hours:
            reasons.append(f"Fits within your {avail_hours}h/week schedule ({item_hours}h required)")
        else:
            reasons.append(f"Requires {item_hours}h/week (exceeds your {avail_hours}h target)")
        if any(s.name.lower() in item_text for s in student_profile.skills):
            reasons.append(f"Leverages your existing skills ({', '.join([s.name for s in student_profile.skills])})")

        return total_score, label, reasons

scoring_service = ScoringService()

File: backend/src/services/scoringService.js
export class ScoringService {
  static calculateScore(item, studentProfile) {
    const stopwords = new Set([
      "i", "want", "to", "become", "a", "an", "the", "in", "and", "have", "know",
      "per", "week", "hours", "hour", "my", "is", "for", "with", "get", "what",
      "which", "how", "can", "help", "related", "explore", "tell", "about"
    ]);

    const goalStr = studentProfile?.goal || "AI Engineer";
    const rawWords = goalStr
      .toLowerCase()
      .split(/\s+/)
      .map((w) => w.replace(/[^a-zA-Z0-9]/g, "").trim())
      .filter((w) => w && !stopwords.has(w) && w.length > 2);

    const goalKeywords = rawWords.length > 0 ? rawWords : ["ai", "engineer"];

    const itemText = (
      `${item.name || ""} ${item.title || ""} ${item.description || ""} ` +
      `${item.skills || ""} ${item.department || ""} ${item.category || ""}`
    ).toLowerCase();

    // 1. Goal Match (55%)
    const goalHits = goalKeywords.filter((kw) => itemText.includes(kw)).length;
    let goalMatch = 10.0;
    if (goalHits > 0) {
      goalMatch = Math.min(100.0, 50.0 + goalHits * 25.0);
    }

    // 2. Skill Match (20%)
    const studentSkills = (studentProfile?.skills || []).map((s) =>
      typeof s === "string" ? s.toLowerCase() : (s.name || "").toLowerCase()
    );
    const itemSkills = (item.skills || "")
      .split("|")
      .map((s) => s.trim().toLowerCase())
      .filter(Boolean);

    let skillMatch = 50.0;
    if (itemSkills.length > 0) {
      const shared = itemSkills.filter((s) => studentSkills.includes(s));
      skillMatch = shared.length > 0
        ? Math.min(100.0, (shared.length / Math.max(1, itemSkills.length)) * 80.0 + 20.0)
        : 40.0;
    }

    // 3. Time Fit (15%)
    const itemHours = parseFloat(item.hours_per_week || 2);
    const availHours = parseFloat(studentProfile?.available_hours_per_week || 6.0);
    let timeFit = 100.0;
    if (itemHours > availHours) {
      const over = itemHours - availHours;
      timeFit = Math.max(10.0, 100.0 - over * 25.0);
    }

    // 4. Opportunity Value (10%)
    const itemType = (item.type || "").toLowerCase();
    let opportunityValue = 75.0;
    if (
      itemType.includes("research") ||
      itemType.includes("fellowship") ||
      itemType.includes("hackathon") ||
      itemType.includes("internship")
    ) {
      opportunityValue = 95.0;
    } else if (itemType.includes("course") || itemType.includes("workshop")) {
      opportunityValue = 85.0;
    }

    // Total Score
    const totalScore = Math.round(
      (goalMatch * 0.55 + skillMatch * 0.20 + timeFit * 0.15 + opportunityValue * 0.10) * 10
    ) / 10;

    // Label
    let label = "Low Match";
    if (totalScore >= 90) label = "Excellent Match";
    else if (totalScore >= 75) label = "Strong Match";
    else if (totalScore >= 60) label = "Good Match";
    else if (totalScore >= 40) label = "Possible Match";

    // Match reasons
    const reasons = [];
    if (goalMatch >= 70) {
      reasons.push(`Directly aligns with your '${goalStr}' goal`);
    }
    if (itemHours <= availHours) {
      reasons.push(`Fits within your ${availHours}h/week schedule (${itemHours}h required)`);
    } else {
      reasons.push(`Requires ${itemHours}h/week (exceeds your ${availHours}h target)`);
    }
    if (studentSkills.some((s) => s && itemText.includes(s))) {
      reasons.push(`Leverages your existing skills (${studentSkills.filter(Boolean).join(", ")})`);
    }

    return {
      score: totalScore,
      label,
      reasons,
    };
  }
}

export const scoringService = ScoringService;
